File: Samples/Objective-C/ConsumerApp/ViewController.m
//
//  ViewController.m
//  ConsumerApp
//
//  Created by Sunku Sneha on 05/03/26.
//
#import "ViewController.h"
#import "NSString+TPSize.h"
#import "TPNetworkManager.h"
#import "TerraPayWalletSDK/TerraPayWalletSDK-Swift.h"
#import "TPPINViewController.h"
#import "UIView+Extension.h"

#pragma mark - UIViewController Toast

@interface UIViewController (TPToast)
- (void)showToastWithMessage:(NSString *)message;
@end

@implementation UIViewController (TPToast)

- (void)showToastWithMessage:(NSString *)message {
//     UIFont *font = [UIFont custom:TPFontWeightRegular size:13.0]; // Example if you have custom font method
    UIFont *font = [UIFont systemFontOfSize:13.0];

    CGSize msgSize = [message tp_sizeForViewWithFont:font
                                               width:CGFLOAT_MAX
                                              height:35.0];
    
    
    CGFloat maxWidth = UIScreen.mainScreen.bounds.size.width - 40.0;
    CGFloat toastWidth = MIN(msgSize.width + 20.0, maxWidth);

    CGFloat x = (self.view.frame.size.width - toastWidth) / 2.0;
    CGFloat y = self.view.frame.size.height - 100.0;

    UILabel *toastLabel = [[UILabel alloc] initWithFrame:CGRectMake(x, y, toastWidth, 35.0)];
    toastLabel.backgroundColor = [UIColor darkGrayColor];

    UIColor *fg = [UIColor colorNamed:@"buttonFG"];
    toastLabel.textColor = fg ?: [UIColor whiteColor];

    toastLabel.textAlignment = NSTextAlignmentCenter;
    toastLabel.font = font;
    toastLabel.text = message;
    toastLabel.alpha = 1.0;
    toastLabel.layer.cornerRadius = 10.0;
    toastLabel.clipsToBounds = YES;

    [self.view addSubview:toastLabel];

    [UIView animateWithDuration:4.0
                          delay:0.1
                        options:UIViewAnimationOptionCurveEaseOut
                     animations:^{
        toastLabel.alpha = 0.0;
    } completion:^(BOOL finished) {
        [toastLabel removeFromSuperview];
    }];
}

@end


@implementation ViewController: UIViewController


#pragma mark - Lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];

    // Initialize constants (equivalent to Swift lets)
    self.subscriberName = @"TerraPay User";
    self.msisdn = @"766901491";
    self.dialCode = @"+256";
    self.balance = @"100000";
    self.currencyCode = @"KES";
    self.countryCode = @"KE";
    self.countryName = @"KENYA";

    [self setUpUIElements];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    // Re-apply corner rounding mask since layer mask depends on bounds
    [self.viewHeader tp_roundCorners:(UIRectCornerBottomLeft | UIRectCornerBottomRight) radius:14.0];
}

#pragma mark - Private

- (void)setUpUIElements {
    [self.viewHeader tp_roundCorners:(UIRectCornerBottomLeft | UIRectCornerBottomRight) radius:14.0];
    [self.viewContainer tp_setRadiusWithShadow];
    [self displayUserInfo];
}

- (void)displayUserInfo {
    self.lblNameCustomer.text = self.subscriberName;
    self.lblNameShort.text = @"TU";
    self.lblMobileNum.text = [NSString stringWithFormat:@"%@%@", self.dialCode, self.msisdn];
    self.lblBalance.text = [NSString stringWithFormat:@"%@ %@", self.currencyCode, self.balance];
}

#pragma mark - Actions

- (IBAction)handleViewTransactionsClicked:(id)sender {
    if (TPNetworkManager.networkStatus != TPNetworkStatusSatisfied) {
        [[TPNetworkManager shared] showNoInternetAlertWithParent:self];
        return;
    }

    // TODO: push/present your transactions screen when network is available.
}

- (IBAction)handleMerchantPaymentsClicked:(id)sender {
  
    TerraPayWalletSDKConfig *config =
        [[TerraPayWalletSDKConfig alloc] initWithController:self
                                          subscriberDialCode:@"+256"
                                          subscriberCountry:@"UG"
                                     subscriberCountryName:@"UGANDA"
                                            subscriberName:@"MTN MoMo"
                                         subscriberMSISDN:@"766901491"
                                       subscriberCurrency:@"UGX"
                                            walletBalance:@"100000"
                                             primaryColor:@"52B44A"
                                           secondaryColor:@"FFFFFF"];
    
    [[TerraPayWalletClient shared] launchWith: config
                            completionHandler:^(TPLaunchType launchType,
                                                TPErrorInfo * _Nullable error,
                                                TPMerchant * _Nullable merchantmodel,
                                                TPPaymentStatus * _Nullable statusModel) {
        
        switch (launchType) {
            case TPLaunchTypeOnPINAuthenticate: {
                if (merchantmodel != nil) {
                    NSLog(@"%@", merchantmodel);
                    [self showTargetHostScreen:merchantmodel];
                }
                break;
            }

            case TPLaunchTypeOnError: {
                if (error.message != nil) {
                    [self showToastWithMessage:error.message];
                }
                break;
            }

            case TPLaunchTypeCancelled: {
                [self showToastWithMessage:@"User cancelled."];
                break;
            }

            case TPLaunchTypeOnPaymentSuccess: {
                NSLog(@"Base app: %@", statusModel.description);
                break;
            }

            case TPLaunchTypeOnPaymentFailure: {
                if (error != nil) {
                    [self showToastWithMessage:error.message];
                } else {
                    [self showToastWithMessage:@"Something went wrong."];
                }
                break;
            }
        }
    }];

    
    
}

#pragma mark - Navigation

- (void)showTargetHostScreen:(TPMerchant *)merchant {
    UIStoryboard *stb = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    UIViewController *vc = [stb instantiateViewControllerWithIdentifier:[TPPINViewController identifier]];
    if ([vc isKindOfClass:[TPPINViewController class]]) {
        TPPINViewController *targetVC = (TPPINViewController *)vc;
        targetVC.modalPresentationStyle = UIModalPresentationFullScreen;
        targetVC.merchantInfo = merchant;
        [self presentViewController:targetVC animated:YES completion:nil];
    }
}

@end

File: Samples/Swift/ConsumerApp/ViewController.swift
//
//  ViewController.swift
//  ConsumerApp
//
//  Created by Sunku Sneha on 18/09/25.
//

import UIKit
import TerraPayWalletSDK

class ViewController: UIViewController {
    
    @IBOutlet weak var viewHeader: UIView!
    @IBOutlet weak var viewContainer: UIView!
    @IBOutlet weak var lblNameCustomer: UILabel!
    @IBOutlet weak var lblNameShort: UILabel!
    @IBOutlet weak var lblMobileNum: UILabel!
    @IBOutlet weak var lblBalance: UILabel!
    
    static let identifier = "TPDashBoardViewController"
    let subscriberName = "TerraPay User"
    let msisdn = "766901491"
    let dialCode = "+256"
    let balance = "100000"
    let currencyCode = "KES"
    let countryCode = "KE"
    let countryName = "KENYA"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setUPUIElements()
    }
    
    private func setUPUIElements() {
        viewHeader.roundCorners(corners: [.bottomLeft, .bottomRight], radius: 14)
        viewContainer.setRadiusWithShadow()
        displayUserInfo()
    }
    
    private func displayUserInfo() {
        lblNameCustomer.text = subscriberName
        lblNameShort.text = "TU"
        lblMobileNum.text = "\(dialCode)\(msisdn)"
        lblBalance.text = "\(currencyCode)" + " " + "\(balance)"
    }
    
    @IBAction func handleViewTransactionsClicked(_ sender: Any) {
        if TPNetworkManager.networkStatus != .satisfied {
            TPNetworkManager.shared.showNoInternetAlert(with: self)
            return
        }
    }
    
    @IBAction func handleMerchantPaymentsClicked(_ sender: Any) {
        
        let config = TerraPayWalletSDKConfig(controller: self,
                                             subscriberDialCode: dialCode,
                                             subscriberCountry: countryCode,
                                             subscriberCountryName: countryName,
                                             subscriberName: subscriberName,
                                             subscriberMSISDN: msisdn,
                                             subscriberCurrency: currencyCode,
                                             walletBalance: balance,
                                             primaryColor: "52B44A",
                                             secondaryColor: "FFFFFF")
        
        TerraPayWalletClient.shared.launch(with: config) { type, error, merchantDetails, status  in
            switch type {
            case .onPINAuthenticate:
                guard let merchantDetails = merchantDetails else { return }
                self.showTargetHostScreen(merchantDetails)
                break
            case .onError:
                if let errmsg = error?.message {
                    self.showToast(message: errmsg)
                }
            case .cancelled:
                self.showToast(message: "User cancelled.")
            case .onPaymentSuccess:
                print("Base app: \(status?.description)")
            case .onPaymentFailure:
                guard let error = error else{
                    self.showToast(message: "Something went wrong.")
                    return
                }
                self.showToast(message: error.message)
            default:
                self.showToast(message: "something went wrong.")
            }
        }
    }
    
    private func showTargetHostScreen(_ merchant: TPMerchant) {
        let stb = UIStoryboard.init(name: "Main", bundle: nil)
        if let targetVC = stb.instantiateViewController(withIdentifier: TPPINViewController.identifier) as? TPPINViewController {
            targetVC.modalPresentationStyle = .fullScreen
            targetVC.merchantInfo = merchant
            present(targetVC, animated: true)
        }
    }
}




extension UIViewController {
    func showToast(message: String) {
        
        var msgSize = message.sizeForView(font: UIFont.custom(.regular, size: 13.0), width: CGFloat.greatestFiniteMagnitude, height: 35)
        msgSize.width += 20
        
        if msgSize.width > UIScreen.main.bounds.width-40 {
            msgSize.width = UIScreen.main.bounds.width-40
        }
        
        let toastLabel = UILabel(frame: CGRect(x: (self.view.frame.size.width-msgSize.width)/2, y: self.view.frame.size.height-100, width: msgSize.width, height: 35))
        toastLabel.backgroundColor = UIColor.darkGray
        toastLabel.textColor = UIColor(named: "buttonFG")
        toastLabel.textAlignment = .center
        toastLabel.font = UIFont.custom(.regular, size: 13.0)
        toastLabel.text = message
        toastLabel.alpha = 1.0
        toastLabel.layer.cornerRadius = 10
        toastLabel.clipsToBounds  =  true
        self.view.addSubview(toastLabel)
        UIView.animate(withDuration: 4.0, delay: 0.1, options: .curveEaseOut, animations: {
            toastLabel.alpha = 0.0
        }, completion: {(_) in
            toastLabel.removeFromSuperview()
        })
    }
}

extension String {
    
    func sizeForView(font: UIFont, width: CGFloat, height: CGFloat) -> CGSize {
        
        let label: UILabel = UILabel(frame: CGRect(x: 0, y: 0, width: width, height: height))
        label.numberOfLines = 0
        label.lineBreakMode = NSLineBreakMode.byWordWrapping
        label.font = font
        label.text = self
        
        label.sizeToFit()
        return label.frame.size
    }
}
